client/core/core.go

```diff
@@ -760,182 +765,140 @@ func (c *Core) Register(form *RegisterForm) error {
-	encKey, err := crypter.Encrypt(privKey.Serialize())
-	if err != nil {
-		return fmt.Errorf("error encrypting private key: %v", err)
+		return err
 	}
 
-	// The account ID is generated from the public key.
-	pubKey := privKey.PubKey()
-
 	// Prepare and sign the registration payload.
+	// The account ID is generated from the public key.
 	dexReg := &msgjson.Register{
-		PubKey: pubKey.Serialize(),
+		PubKey: privKey.PubKey().Serialize(),
 		Time:   encode.UnixMilliU(time.Now()),
 	}
-	err = sign(privKey, dexReg)
-	if err != nil {
-		return err
-	}
-
-	// Create and send the the request, grabbing the result in the callback.
-	regMsg, err := msgjson.NewRequest(dc.NextID(), msgjson.RegisterRoute, dexReg)
-	if err != nil {
-		return fmt.Errorf("error encoding message: %v", err)
-	}
 	regRes := new(msgjson.RegisterResult)
-	errChan := make(chan error, 1)
-	err = dc.Request(regMsg, func(msg *msgjson.Message) {
-		errChan <- msg.UnmarshalResult(regRes)
-	})
+	err = signAndRequest(dexConn.WsConn, msgjson.RegisterRoute, dexReg, privKey, regRes)
 	if err != nil {
-		return fmt.Errorf("'register' request error: %v", err)
-	}
-	err = extractError(errChan, requestTimeout, "register")
-	if err != nil {
-		return fmt.Errorf("'register' result decode error: %v", err)
+		return err
 	}
 
-	// Check the server's signature.
+	// Check the DEX server's signature.
 	msg, err := regRes.Serialize()
 	if err != nil {
 		return fmt.Errorf("error serializing RegisterResult for signature check: %v", err)
 	}
-
-	// Create a public key for this account.
 	dexPubKey, err := checkSigS256(msg, regRes.DEXPubKey, regRes.Sig)
 	if err != nil {
 		return fmt.Errorf("DEX signature validation error: %v", err)
 	}
 
 	// Check that the fee is non-zero.
 	if regRes.Fee == 0 {
-		return fmt.Errorf("zero registration fees not supported")
+		return fmt.Errorf("zero registration fees not allowed")
 	}
-	if regRes.Fee != dc.cfg.Fee {
-		return fmt.Errorf("DEX 'register' result fee doesn't match the 'config' value. %d != %d", regRes.Fee, dc.cfg.Fee)
+	if regRes.Fee != dexConn.cfg.Fee {
+		return fmt.Errorf("DEX 'register' result fee doesn't match the 'config' value. %d != %d", regRes.Fee, dexConn.cfg.Fee)
 	}
 	if regRes.Fee != form.Fee {
 		return fmt.Errorf("registration fee provided to Register does not match the DEX registration fee. %d != %d", form.Fee, regRes.Fee)
 	}
 
+	// Server's signature and reg fee checks out. Save the new dex account info
+	// before paying fee, in case of db save errors.
+	dexConn.acct.dexPubKey = dexPubKey
+	err = c.db.CreateAccount(dexConn.acct.dbInfo())
+	if err != nil {
+		return fmt.Errorf("error saving account: %v", err)
+	}
+
 	// Pay the registration fee.
 	log.Infof("Attempting registration fee payment for %s of %d units of %s", regRes.Address,
 		regRes.Fee, regAsset.Symbol)
 	coin, err := wallet.PayFee(regRes.Address, regRes.Fee, regAsset)
 	if err != nil {
 		return fmt.Errorf("error paying registration fee: %v", err)
 	}
```

Comment:
Because storing an account without a fee coin is the same as not storing an account at all (see [here](https://github.com/itswisdomagain/dcrdex/blob/a20e6b6da8fb710f9c70236998ffd14b27a4d894/client/core/core.go#L1526)), the order here doesn't matter. This just adds an additional update. Old way is simpler. 

I do like the separate functions for `db.CreateAccount` and `db.UpdateAccount`.

Reply:
Can we rethink that? What motivated that change is the known (although unexpected) possibility of a user registering with a DEX server and paying fees but losing access when dexc is restarted because the user's account ID was not saved to db.

Instead of discarding accounts without feeCoin info when dexc restarts, we should attempt to recover the account. Perhaps attempt to connect to the server (which I expect would fail if the server has not previously been notified of fee payment). If the connection succeeds, we've technically recovered the account, although we'd need to set the feeCoin info in db to complete the account recovery.

To avoid unnecessary account recovery attempts, we can delete the account info from db if fee payment fails. We can also retry to update the dbAccount feeCoin property when the fee tx get the required nConfs.

The only other time account recovery will fail is if the fee was paid but the server was not notified prolly because the required confs were not attained before dexc was shutdown or some other error occurred.

If https://github.com/decred/dcrdex/issues/306#issuecomment-622995854 is considered, then we can effectively eliminate cases where dex server connection fails after a user has paid because once fee payment tx is broadcasted, the dex server would have knowledge of the payment instead of waiting for `regFeeConfs` before notifying the server of payment.

Ultimately however, if automatic account recovery fails, manual account recovery can be attempted with the account info previously saved to db; whereas if there was no initial save of account info before paying fees, the account would be lost for life if the account info could not be saved after fee payment and dexc restarts.

Reply:
> Instead of discarding accounts without feeCoin info when dexc restarts, we should attempt to recover the account.

Recover what account? The DEX has no record of the account. The "account" you're talking about is just a random private key that's never been used. There's no reason to worry about "recovering" it. 

> To avoid unnecessary account recovery attempts, we can delete the account info from db if fee payment fails.

Then what's the point in saving it?

> If #306 (comment) is considered, then we can effectively eliminate cases where dex server connection fails after a user has paid because once fee payment tx is broadcasted

That was considered. Have you read the spec? User pays after the fee has requisite confirmations. 

> If #306 (comment) is considered, then we can effective

Reply:
> > Instead of discarding accounts without feeCoin info when dexc restarts, we should attempt to recover the account.
> 
> Recover what account? The DEX has no record of the account. The "account" you're talking about is just a random private key that's never been used. There's no reason to worry about "recovering" it.

In a scenario where the user successfully registered with the DEX server, does not the server store the account info for the newly registered user?

With the current implementation, **if** dex server registration succeeds, fee is paid but saving account info to db fails, the user has to connect to the dex server again and pay another fee.

The initial account with the dex server is completely lost because the private key and account id were not saved.

Reply:
If storage of the reg fee coin ID fails, **Core** never gets to `notifyfee`, and the server has no idea. 

This is likely related to the spec mistake that you discovered, which seemed to imply an additional notification immediately after the fee is paid, but before waiting for confirmations. That was a mistake. Sorry about that. The current implementation does not send an earlier notification to the server, just `notifyfee` after confirmations.

Reply:
After reading your issue and reading through this again, I think maybe you're right that it does make sense to store the account private key before the fee coin is requested. Also, I forgot the account key is linked to the registration fee address through the `register` request, so we shouldn't toss the key. My bad.